File: crypto/src/main/java/com/blockstream/lightning/LightningBridge.kt
package com.blockstream.lightning

import android.util.Base64
import breez_sdk.BlockingBreezServices
import breez_sdk.BreezEvent
import breez_sdk.Config
import breez_sdk.EnvironmentType
import breez_sdk.EventListener
import breez_sdk.GreenlightCredentials
import breez_sdk.GreenlightNodeConfig
import breez_sdk.InputType
import breez_sdk.LnInvoice
import breez_sdk.LnUrlAuthRequestData
import breez_sdk.LnUrlCallbackStatus
import breez_sdk.LnUrlPayRequestData
import breez_sdk.LnUrlPayResult
import breez_sdk.LnUrlWithdrawRequestData
import breez_sdk.LspInformation
import breez_sdk.Network
import breez_sdk.NodeConfig
import breez_sdk.NodeState
import breez_sdk.Payment
import breez_sdk.PaymentTypeFilter
import breez_sdk.RecommendedFees
import breez_sdk.SwapInfo
import breez_sdk.connect
import breez_sdk.defaultConfig
import breez_sdk.mnemonicToSeed
import breez_sdk.parseInput
import breez_sdk.parseInvoice
import com.blockstream.common.lightning.AppGreenlightCredentials
import com.blockstream.crypto.BuildConfig
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.datetime.Clock
import mu.KLogging
import java.io.File

class LightningBridge constructor(
    val workingDir: File
) : EventListener {

    private var breezSdkOrNull: BlockingBreezServices? = null
    private val breezSdk
        get() = breezSdkOrNull!!
    private val network
        get() = Network.BITCOIN

    // Emulate GreenlightCredentials
    var appGreenlightCredentials: AppGreenlightCredentials? = null

    private val _nodeInfoStateFlow = MutableStateFlow<NodeState>(
        NodeState(
            id = "",
            blockHeight = 0u,
            channelsBalanceMsat = 0u,
            onchainBalanceMsat = 0u,
            utxos = listOf(),
            maxPayableMsat = 0u,
            maxReceivableMsat = 0u,
            maxSinglePaymentAmountMsat = 0u,
            maxChanReserveMsats = 0u,
            connectedPeers = listOf(),
            inboundLiquidityMsats = 0u
        )
    )

    val nodeInfoStateFlow
        get() = _nodeInfoStateFlow.asStateFlow()

    private val _eventSharedFlow = MutableSharedFlow<BreezEvent>(extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)

    val eventSharedFlow
        get() = _eventSharedFlow.asSharedFlow()

    private val _lspInfoStateFlow = MutableStateFlow<LspInformation?>(null)

    val lspInfoStateFlow
        get() = _lspInfoStateFlow.asStateFlow()

    private val _swapInfoStateFlow = MutableStateFlow<List<Pair<SwapInfo, Boolean>>>(listOf())

    val swapInfoStateFlow
        get() = _swapInfoStateFlow.asStateFlow()

    private fun createConfig(partnerCredentials: GreenlightCredentials?): Config {

        val apiKey = BuildConfig.BREEZ_API_KEY.let { base64 ->
            try{
                String(Base64.decode(base64, Base64.DEFAULT))
            }catch (e: Exception){
                ""
            }
        }

        val nodeConfig = NodeConfig.Greenlight(config = GreenlightNodeConfig(partnerCredentials = partnerCredentials, inviteCode = null))

        return defaultConfig(EnvironmentType.PRODUCTION, apiKey, nodeConfig).also {
            it.workingDir = workingDir.absolutePath
        }
    }

    private var inputTypeCache: Pair<String, InputType>? = null

    var isConnected = false
        private set

    init {
        logger.info { "Lightning SDK: ${workingDir.absoluteFile}" }

        if (!workingDir.exists()) {
            workingDir.mkdirs()
        }
    }

    private val greenlightCredentials by lazy {
        if (BuildConfig.GREENLIGHT_DEVICE_CERT.isNotBlank() && BuildConfig.GREENLIGHT_DEVICE_KEY.isNotBlank()) {
            GreenlightCredentials(
                deviceKey = Base64.decode(BuildConfig.GREENLIGHT_DEVICE_KEY, Base64.DEFAULT)
                    .toUByteArray().toTypedArray().toList(),
                deviceCert = Base64.decode(BuildConfig.GREENLIGHT_DEVICE_CERT, Base64.DEFAULT)
                    .toUByteArray().toTypedArray().toList(),
            )
        } else null
    }

    fun connectToGreenlight(mnemonic: String, isRestore: Boolean): Boolean {
        val partnerCredentials = if (isRestore) null else greenlightCredentials
        return start(mnemonic, partnerCredentials)
    }

    private fun start(mnemonic: String, partnerCredentials: GreenlightCredentials?): Boolean {
        if (breezSdkOrNull != null) {
            return true
        }

        try{
            breezSdkOrNull = connect(
                config = createConfig(partnerCredentials),
                seed = mnemonicToSeed(mnemonic),
                listener = this
            )

            appGreenlightCredentials = greenlightCredentials?.let {
                AppGreenlightCredentials.fromGreenlightCredentials(
                    it
                )
            }

            isConnected = true

            updateNodeInfo()

            updateLspInformation()

            return true
        }catch (e: Exception){
            e.printStackTrace()
        }
        return false
    }

    private fun updateLspInformation() {
        try {
            breezSdk.lspId()?.also {
                _lspInfoStateFlow.tryEmit(breezSdk.fetchLspInfo(it).also {
                    logger.debug { "LspInformation: $it" }
                })
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun updateNodeInfo(): NodeState? {
        return breezSdk.nodeInfo()?.also {
            logger.debug { "NodeState: $it" }
            _nodeInfoStateFlow.value = it
            // Emit New Block as a way to identify established connection
            _eventSharedFlow.tryEmit(BreezEvent.NewBlock(it.blockHeight))
        }
    }

    fun balance(): Long? {
        if(breezSdkOrNull == null){
            return null
        }
        return updateNodeInfo()?.also {
            logger.debug { "Balance: ${it.channelsBalanceSatoshi()}" }
        }?.channelsBalanceSatoshi()
    }

    fun parseBolt11(bolt11: String): LnInvoice? {
        return try {
            if (bolt11.isNotBlank()) {
                parseInvoice(bolt11)
            } else {
                null
            }
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    // Cache input type to avoid server communication (LNURL) and propably a change in the requested data
    fun parseBoltOrLNUrlAndCache(input: String): InputType? {
        return (inputTypeCache.takeIf { inputTypeCache?.first == input }?.second
            ?: parseBoltOrLNUrl(input))
    }

    private fun parseBoltOrLNUrl(input: String?): InputType? {
        return try {
            if (!input.isNullOrBlank()) {
                parseInput(input).takeIf {
                    it is InputType.Bolt11 || it is InputType.LnUrlPay || it is InputType.LnUrlWithdraw || it is InputType.LnUrlAuth
                }?.also {
                    inputTypeCache = input to it
                }
            } else {
                null
            }
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    fun getTransactions(): List<Payment>? {
        if(breezSdkOrNull == null){
            return null
        }

        return breezSdk.listPayments(
            filter = PaymentTypeFilter.ALL,
            fromTimestamp = null,
            toTimestamp = null
        ).onEach {
            logger.info { it }
        }.also {
            updateSwapInfo()
        }
    }

    fun createInvoice(satoshi: Long, description: String): LnInvoice {
        return try {
            breezSdk.receivePayment(amountSats = satoshi.toULong(), description = description)
        } catch (e: Exception) {
            throw exceptionWithNodeId(e)
        }
    }

    fun receiveOnchain(): SwapInfo? {
        return try {
            breezSdk.receiveOnchain()
        } catch (e: Exception) {
            e.printStackTrace()

            if (e.message?.contains(
                    "Swap in progress was detected".lowercase(),
                    ignoreCase = true
                ) == true
            ) {
                throw Exception("id_swap_in_progress_was_detected")
            }

            throw exceptionWithNodeId(e)
        }
    }

    private fun updateSwapInfo() {
        _swapInfoStateFlow.value = (listOfNotNull(breezSdk.inProgressSwap()).map {
            it to false
        } + breezSdk.listRefundables().map {
            it to true
        })
    }

    fun refund(swapAddress: String, toAddress: String, satPerVbyte: UInt?): String? {
        return try {
            breezSdk.refund(
                swapAddress = swapAddress,
                toAddress = toAddress,
                satPerVbyte = satPerVbyte ?: breezSdk.recommendedFees().economyFee.toUInt()
            )
        } catch (e: Exception) {
            e.printStackTrace()
            throw exceptionWithNodeId(e)
        } finally {
            updateSwapInfo()
        }
    }

    fun sweep(toAddress: String, satPerVbyte: UInt?){
        return try {
            breezSdk.sweep(
                toAddress = toAddress,
                feeRateSatsPerVbyte = satPerVbyte?.toULong() ?: breezSdk.recommendedFees().economyFee.toULong()
            )
        } catch (e: Exception) {
            e.printStackTrace()
            throw exceptionWithNodeId(e)
        } finally {
            updateNodeInfo()
        }
    }

    fun recommendedFees(): RecommendedFees {
        return breezSdk.recommendedFees()
    }

    fun sendPayment(bolt11: String, satoshi: Long?): Payment {
        return try {
            breezSdk.sendPayment(bolt11, amountSats = satoshi?.toULong())
        } catch (e: Exception) {
            throw exceptionWithNodeId(e)
        }
    }

    fun payLnUrl(requestData: LnUrlPayRequestData, amount: Long, comment: String): LnUrlPayResult {
        return try {
            breezSdk.payLnurl(requestData, amount.toULong(), comment)
        } catch (e: Exception) {
            throw exceptionWithNodeId(e)
        }
    }

    fun authLnUrl(requestData: LnUrlAuthRequestData): LnUrlCallbackStatus {
        return try {
            breezSdk.lnurlAuth(requestData)
        } catch (e: Exception) {
            throw exceptionWithNodeId(e)
        }
    }

    fun withdrawLnurl(
        requestData: LnUrlWithdrawRequestData,
        amount: Long,
        description: String?
    ): LnUrlCallbackStatus {
        return try {
            breezSdk.withdrawLnurl(requestData, amount.toULong(), description)
        } catch (e: Exception) {
            throw exceptionWithNodeId(e)
        }
    }

    fun listLisps(): List<LspInformation> {
        return breezSdk.listLsps()
    }

    fun connectLsp(id: String) {
        breezSdk.connectLsp(lspId = id)
    }

    fun lspId(): String? {
        return breezSdk.lspId()
    }

    fun fetchLspInfo(id: String): LspInformation? {
        return breezSdkOrNull?.fetchLspInfo(id)
    }

    fun closeLspChannels(){
        breezSdkOrNull?.closeLspChannels()
    }

    fun stop() {
        try {
            breezSdkOrNull?.disconnect()
            breezSdkOrNull = null
        } catch (e: Exception) {
            e.printStackTrace()
        }
        breezSdkOrNull?.close()
    }

    override fun onEvent(e: BreezEvent) {
        logger.info { "Breez onEvent $e" }
        _eventSharedFlow.tryEmit(e)
    }

    private fun exceptionWithNodeId(exception: Exception) =
        Exception("${exception.message}\nNodeId: ${_nodeInfoStateFlow.value.id}\nTimestamp: ${Clock.System.now().epochSeconds}", exception.cause)

    companion object : KLogging()
}

File: lightning/LightningBridge.swift
import Foundation
import BreezSDK

public typealias Long = UInt64

public protocol LightningEventListener {
    func onLightningEvent(event: BreezEvent)
}

class LogStreamListener: LogStream {
    func log(l: LogEntry){
      print("BREEZ: \(l.line)");
    }
}

public class LightningBridge {

    let testnet: Bool
    public var appGreenlightCredentials: AppGreenlightCredentials?
    var breezSdk: BlockingBreezServices?
    var eventListener: LightningEventListener
    var workingDir: URL
    private var network: Network { testnet ? .testnet : .bitcoin }
    private var environment: EnvironmentType { testnet ? .staging : .production }

    static public var BREEZ_API_KEY: String? {
        let content = Bundle.main.infoDictionary?["BREEZ_API_KEY"] as? String
        // print("BREEZ_API_KEY: \(content)")
        if content == nil { print("BREEZ_API_KEY: UNDEFINED") }
        return content
    }
    static public var GREENLIGHT_DEVICE_CERT: Data? {
        let path = Bundle.main.path(forResource: "green", ofType: "crt") ?? ""
        let content = try? String(contentsOf: URL(fileURLWithPath: path), encoding: .utf8)
        if let content = content?.filter({ !$0.isWhitespace }) {
            // print("GREENLIGHT_DEVICE_CERT: \(content)")
            return Data(base64Encoded: content)
        } else {
            print("GREENLIGHT_DEVICE_CERT: UNDEFINED")
        }
        return nil
    }
    static public var GREENLIGHT_DEVICE_KEY: Data? {
        let path = Bundle.main.path(forResource: "green", ofType: "pem") ?? ""
        let content = try? String(contentsOf: URL(fileURLWithPath: path), encoding: .utf8)
        if let content = content?.filter({ !$0.isWhitespace }) {
            // print("GREENLIGHT_DEVICE_KEY: \(content)")
            return Data(base64Encoded: content)
        } else {
            print("GREENLIGHT_DEVICE_KEY: UNDEFINED")
        }
        return nil
    }
    static var CREDENTIALS: GreenlightCredentials? {
        if let cert = LightningBridge.GREENLIGHT_DEVICE_CERT,
           let key = LightningBridge.GREENLIGHT_DEVICE_KEY {
            return GreenlightCredentials(deviceKey: [UInt8](key), deviceCert: [UInt8](cert))
        }
        return nil
    }

    public init(testnet: Bool,
                workingDir: URL,
                eventListener: LightningEventListener) {
        self.testnet = testnet
        self.eventListener = eventListener
        self.workingDir = workingDir
    }

    public static func configure() {
        try? setLogStream(logStream: LogStreamListener())
    }

    public func connectToGreenlight(mnemonic: String, isRestore: Bool) -> Bool {
        let partnerCredentials = isRestore ? nil : LightningBridge.CREDENTIALS
        return start(mnemonic: mnemonic, partnerCredentials: partnerCredentials)
    }

    private func createConfig(_ partnerCredentials: GreenlightCredentials?) -> Config {
        let greenlightConfig = GreenlightNodeConfig(partnerCredentials: partnerCredentials, inviteCode: nil)
        let nodeConfig = NodeConfig.greenlight(config: greenlightConfig)
        var config = defaultConfig(envType: environment,
                      apiKey: LightningBridge.BREEZ_API_KEY ?? "",
                      nodeConfig: nodeConfig)
        config.workingDir = workingDir.path
        try? FileManager.default.createDirectory(atPath: workingDir.path, withIntermediateDirectories: true)
        return config
    }

    private func start(mnemonic: String, partnerCredentials: GreenlightCredentials?) -> Bool {
        if breezSdk != nil {
            return true
        }
        breezSdk = try? connect(config: createConfig(partnerCredentials),
                seed: mnemonicToSeed(phrase: mnemonic),
                listener: self)
        if breezSdk == nil {
            return false
        }
        
        if let credentials = LightningBridge.CREDENTIALS {
            appGreenlightCredentials = AppGreenlightCredentials(gc: credentials)
        }
        _ = updateNodeInfo()
        _ = updateLspInformation()
        return true
    }
    
    public func stop(){
        try? breezSdk?.disconnect()
        breezSdk = nil
    }

    public func updateLspInformation() -> LspInformation? {
        if let id = try? breezSdk?.lspId() {
            return try? breezSdk?.fetchLspInfo(lspId: id)
        }
        return nil
    }

    public func updateNodeInfo() -> NodeState? {
        let res = try? breezSdk?.nodeInfo()
        print ("NodeInfo: \(res.debugDescription)")
        return res
    }

    public func balance() -> UInt64? {
        return updateNodeInfo()?.channelsBalanceSatoshi
    }

    public func parseBolt11(bolt11: String) -> LnInvoice? {
        print ("Parse invoice: \(bolt11)")
        if bolt11.isEmpty { return nil }
        do { return try parseInvoice(invoice: bolt11) }
        catch { print ("Parse invoice: \(error.localizedDescription)"); return nil }
    }

    public func parseBoltOrLNUrl(input: String?) -> InputType? {
        print("Parse input: \(input)")
        guard let input = input else { return nil }
        return try? parseInput(s: input)
    }

    public func getTransactions() -> [Payment] {
        let list = try? breezSdk?.listPayments(filter: PaymentTypeFilter.all, fromTimestamp: nil, toTimestamp: nil)
        list?.forEach { print("Payment: \($0)") }
        return list ?? []
    }

    public func createInvoice(satoshi: Long, description: String) throws -> LnInvoice? {
        let payment = try breezSdk?.receivePayment(amountSats: satoshi, description: description)
        print("createInvoice \(payment)")
        return payment
    }
    
    public func refund(swapAddress: String, toAddress: String, satPerVbyte: UInt32?) -> String? {
        let refund = try? breezSdk?.refund(swapAddress: swapAddress, toAddress: toAddress, satPerVbyte: satPerVbyte ?? UInt32(recommendedFees()?.economyFee ?? 0))
        print("refund \(refund)")
        return refund
    }

    public func swapProgress() -> SwapInfo? {
        try? breezSdk?.inProgressSwap()
    }

    public func swapList() -> [SwapInfo]? {
        try? breezSdk?.listRefundables()
    }

    public func receiveOnchain() -> SwapInfo? {
        return try? breezSdk?.receiveOnchain()
    }

    public func recommendedFees() -> RecommendedFees? {
        return try? breezSdk?.recommendedFees()
    }

    public func sendPayment(bolt11: String, satoshi: UInt64? = nil) throws -> Payment? {
        return try breezSdk?.sendPayment(bolt11: bolt11, amountSats: satoshi)
    }

    public func payLnUrl(requestData: LnUrlPayRequestData, amount: Long, comment: String) -> LnUrlPayResult? {
        return try? breezSdk?.payLnurl(reqData: requestData, amountSats: amount, comment: comment)
    }

    public func authLnUrl(requestData: LnUrlAuthRequestData) -> LnUrlCallbackStatus? {
        return try? breezSdk?.lnurlAuth(reqData: requestData)
    }

    public func withdrawLnurl(requestData: LnUrlWithdrawRequestData, amount: Long, description: String?) -> LnUrlCallbackStatus? {
        return try? breezSdk?.withdrawLnurl(reqData: requestData, amountSats: amount, description: description)
    }

    public func listLisps() -> [LspInformation]? {
        return try? breezSdk?.listLsps()
    }

    public func connectLsp(id: String) {
        try? breezSdk?.connectLsp(lspId: id)
    }

    public func lspId() -> String? {
        return try? breezSdk?.lspId()
    }

    public func fetchLspInfo(id: String) -> LspInformation? {
        return try? breezSdk?.fetchLspInfo(lspId: id)
    }
}

extension LightningBridge: EventListener {
    public func onEvent(e: BreezEvent) {
        print ("Breez onEvent: \(e)")
        eventListener.onLightningEvent(event: e)
        switch e {
        case BreezEvent.invoicePaid( _):
            break
        case BreezEvent.paymentSucceed(_):
            break
        default:
            break
        }
    }
}

